src/storage/src/render/upsert.rs

```diff
@@ -203,6 +213,20 @@ where
                     // Find the prefix that we can emit
                     let idx = stash.partition_point(|(ts, _, _, _)| !upper.less_equal(ts));
 
+                    // Read the previous values _per key_ out of `state`, recording it
+                    // along with the value with the _latest timestamp for that key_.
+                    commands_state.clear();
+                    for (_, key, _, _) in stash.iter().take(idx) {
+                        let state_value = state.get(key).cloned();
+                        commands_state.insert(
+                            *key,
+                            StateValue {
```

Comment:
I think this can be just `Some(state_value)`. We know for a fact that all the keys in `commands_state` will be modified since the only reason they are here is because there is a command for them. We can use `None` as a tombstone so that we can commit the deletions at the end.

Reply:
This is a good point